smarthome.go

```diff
@@ -1,118 +1,204 @@
 package gbridge
 
 import (
-	"log"
-	"net/http"
 	"encoding/json"
-	"os"
+	"errors"
+	"github.com/pborges/gbridge/oauth"
+	"github.com/pborges/gbridge/proto"
 	"io"
+	"log"
+	"net/http"
+	"sync"
 )
 
-func (b *Bridge) HandleSmartHome(w http.ResponseWriter, r *http.Request) {
-	log.Println(r.Method, r.RequestURI)
+//todo: naive concurrency scheme, this should be looked at and redone
 
-	w.Header().Add("Access-Control-Allow-Origin", "*")
-	w.Header().Add("Access-Control-Allow-Headers", "Content-Type, Authorization")
+type SmartHome struct {
+	agents map[string]agentContext
+	lock   sync.RWMutex
+}
+
+func (s *SmartHome) decodeAndHandle(agent agentContext, r io.Reader) proto.IntentMessageResponse {
+	req := proto.IntentMessageRequest{}
 
-	req := IntentMessageRequest{}
-	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+	if err := json.NewDecoder(r).Decode(&req); err != nil {
 		log.Println("error decoding request:", err)
 	}
+
+	res := proto.IntentMessageResponse{
+		RequestId: req.RequestId,
+	}
+
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
 	for _, i := range req.Inputs {
-		log.Printf("Intent: %s-> %s\n", i.Intent, string(i.Payload))
+		log.Printf("[%s] Intent: %s-> %s\n", agent.AgentUserId, i.Intent, string(i.Payload))
 		switch i.Intent {
-		case "action.devices.QUERY":
-			requestBody := QueryRequest{}
-			if err := json.Unmarshal(i.Payload, &requestBody); err != nil {
-				log.Println(err)
-				return
+		case "action.devices.SYNC":
+			res.Payload = s.handleSyncIntent(agent)
+		case "action.devices.EXECUTE":
+			requestBody := proto.ExecRequest{}
+			if err := json.Unmarshal(i.Payload, &requestBody); err == nil {
+				res.Payload = s.handleExecuteIntent(agent, requestBody)
+			} else {
+				res.Payload = proto.ErrorResponse{
+					Status:    proto.CommandStatusError,
+					ErrorCode: proto.ErrorCodeProtocolError.Error(),
+				}
 			}
-			log.Printf("QUERY: %+v\n", requestBody)
+		case "action.devices.QUERY":
+			//todo handle query
+		}
+	}
+	return res
+}
 
-			responseBody := QueryResponse{
-				Devices: make(map[string]DeviceState),
+func (s *SmartHome) Handle() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+
+		if agent, ok := s.agents[oauth.GetAgentUserIdFromHeader(r)]; ok {
+			res := s.decodeAndHandle(agent, r.Body)
+			if err := json.NewEncoder(io.MultiWriter(w, log.Writer())).Encode(res); err != nil {
+				log.Println("error encoding sync response:", err)
+				http.Error(w, "error encoding sync response: "+err.Error(), http.StatusInternalServerError)
 			}
+		} else {
+			http.Error(w, "unable to find agentUserId", http.StatusInternalServerError)
```

Comment:
This is a bug, SmartHome.Handle() should not return an error, at this point it has "passed" Agent authentication and SmartHome.Handle() needs only return an empty list of devices if it cant find an agent for whatever reason

Reply:
Confirming this bug.